File: R/gh_top_lang_card.R

# Top Language Card ---------------------------------------------------

#' Create Github Top Languages Card
#'
#' @param username Your Github username
#' @param exclude_repo if not `NULL`, specify character vector of repo name(s) to exclude individual repositories.
#' @param hide if not `NULL`, specify character vector of language name(s) to hide
#' @param langs_count Option to specify the number of languages shown on the card, must be an integer from 1 to 10, default is 5.
#' @param layout_compact if `TRUE`, card will show in a compact layout
#' @param theme If not `NULL`, provide build in theme names
#'
#' @return A character rendering GitHub Top Languages Card to be inserted into R Markdown "github_document"
#' @export
#' @source https://github.com/anuraghazra/github-readme-stats
#' 
gh_top_lang_card <- function(username,
                             exclude_repo = NULL,
                             hide = NULL,
                             langs_count = 5,
                             layout_compact = FALSE,
                             theme = NULL
){
  
  url <- gh_top_lang_card_url(username = username,
                              exclude_repo = exclude_repo, 
                              hide = hide, 
                              langs_count = langs_count, 
                              layout_compact = layout_compact)
  
  # Build URL
  glue::glue("[![Top Langs]({url})](https://github.com/{username}/{username})
")
  
}



# Top Language Card URL ---------------------------------------------------

#' Create Github Top Languages Card URL
#'
#' @param username Your Github username
#' @param exclude_repo if not `NULL`, specify character vector of repo name(s) to exclude individual repositories.
#' @param hide if not `NULL`, specify character vector of language name(s) to hide
#' @param langs_count Option to specify the number of languages shown on the card, must be an integer from 1 to 10, default is 5.
#' @param layout_compact if `TRUE`, card will show in a compact layout
#' @param theme If not `NULL`, provide build in theme names
#'
#' @return A URL of GitHub Top Languages Card 
#' @export
#' @source https://github.com/anuraghazra/github-readme-stats
#' 
gh_top_lang_card_url <- function(username,
                                 exclude_repo = NULL,
                                 hide = NULL,
                                 langs_count = 5,
                                 layout_compact = FALSE,
                                 theme = NULL
){
  
  ## Exclude Individual Repo
  exclude_repo <- if (is.null(exclude_repo)) {
    ""
  } else {
    if (!is.character(exclude_repo)) stop("`exclude_repo` must be a character indicate repo names to exclude.", call. = FALSE)
    paste0("&exclude_repo=", paste(exclude_repo, collapse = ","))
  }
  
  ## Hide Individual Language
  hide <- if (is.null(hide)) {
    ""
  } else {
    if (!is.character(hide)) stop("`hide` must be a character indicate which language(s) to hide.")
    paste0("&hide=", paste(hide, collapse = ","))
  }
  
  ## Language Count
  valid_langs_count <- (length(langs_count) == 1) && (langs_count %in% 1:10)
  if(!valid_langs_count) stop("`langs_count` must be one of 1 to 10.")
  
  langs_count <- paste0("&langs_count=", langs_count)
  
  ## Compact Language Card Layout
  layout_compact <- if(layout_compact){
    "&layout=compact"
  } else {
    ""
  }
  
  ## Theme
  theme <- if (is.null(theme)) {
    ""
  } else {
    ## Choose Theme
    paste0("&theme=", theme)
  }
  
  
  # Build URL
  glue::glue("https://github-readme-stats.vercel.app/api/top-langs/?username={username}{exclude_repo}{hide}{langs_count}{layout_compact}{theme}")
  
}

File: src/gh_top_lang_card.py
"""GitHub Top Languages Card URL generation."""


def gh_top_lang_card_url(
    username: str,
    exclude_repo: list[str] | None = None,
    hide: list[str] | None = None,
    langs_count: int = 5,
    layout_compact: bool = False,
    theme: str | None = None,
) -> str:
    """
    Create GitHub Top Languages Card URL.

    Parameters
    ----------
    username : str
        Your GitHub username
    exclude_repo : list[str] | None, optional
        Repository names to exclude from language statistics, by default None
    hide : list[str] | None, optional
        Language names to hide (e.g., ["html", "scss"]), by default None
    langs_count : int, optional
        Number of languages to display (must be between 1 and 10), by default 5
    layout_compact : bool, optional
        Use compact card layout, by default False
    theme : str | None, optional
        Built-in theme name, by default None

    Returns
    -------
    str
        URL of GitHub Top Languages Card

    Raises
    ------
    TypeError
        If exclude_repo or hide are not lists of strings
    ValueError
        If langs_count is not between 1 and 10

    Notes
    -----
    Based on: https://github.com/anuraghazra/github-readme-stats

    Examples
    --------
    >>> gh_top_lang_card_url("octocat")
    'https://github-readme-stats.vercel.app/api/top-langs/?username=octocat&langs_count=5'

    >>> gh_top_lang_card_url("octocat", hide=["html"], layout_compact=True, theme="algolia")
    'https://github-readme-stats.vercel.app/api/top-langs/?username=octocat&hide=html&langs_count=5&layout=compact&theme=algolia'
    """
    # Validate exclude_repo
    if exclude_repo is not None:
        if not isinstance(exclude_repo, list) or not all(
            isinstance(repo, str) for repo in exclude_repo
        ):
            raise TypeError("`exclude_repo` must be a list of repository names to exclude")

    # Validate hide
    if hide is not None:
        if not isinstance(hide, list) or not all(isinstance(lang, str) for lang in hide):
            raise TypeError("`hide` must be a list of language names to hide")

    # Validate langs_count
    if not isinstance(langs_count, int) or not (1 <= langs_count <= 10):
        raise ValueError("`langs_count` must be an integer between 1 and 10")

    # Build query parameters
    params = []

    if exclude_repo:
        exclude_str = ",".join(exclude_repo)
        params.append(f"exclude_repo={exclude_str}")

    if hide:
        hide_str = ",".join(hide)
        params.append(f"hide={hide_str}")

    params.append(f"langs_count={langs_count}")

    if layout_compact:
        params.append("layout=compact")

    if theme:
        params.append(f"theme={theme}")

    # Construct URL
    base_url = f"https://github-readme-stats.vercel.app/api/top-langs/?username={username}"
    query_string = "&".join(params)

    return f"{base_url}&{query_string}"


def gh_top_lang_card(
    username: str,
    exclude_repo: list[str] | None = None,
    hide: list[str] | None = None,
    langs_count: int = 5,
    layout_compact: bool = False,
    theme: str | None = None,
) -> str:
    """
    Create GitHub Top Languages Card markdown link.

    Parameters
    ----------
    username : str
        Your GitHub username
    exclude_repo : list[str] | None, optional
        Repository names to exclude, by default None
    hide : list[str] | None, optional
        Language names to hide, by default None
    langs_count : int, optional
        Number of languages to display (1-10), by default 5
    layout_compact : bool, optional
        Use compact layout, by default False
    theme : str | None, optional
        Built-in theme name, by default None

    Returns
    -------
    str
        Markdown link rendering GitHub Top Languages Card

    Notes
    -----
    Based on: https://github.com/anuraghazra/github-readme-stats
    """
    url = gh_top_lang_card_url(
        username=username,
        exclude_repo=exclude_repo,
        hide=hide,
        langs_count=langs_count,
        layout_compact=layout_compact,
        theme=theme,
    )

    return f"[![Top Langs]({url})](https://github.com/{username}/{username})\n"
